### PR TITLE
Add new SQLCheckOperators

### LINKED ISSUE #23874: Add Templated SQL Check Operators for Columns and Tables
### Description

The new SQL Check Operators would provide templated quality checks so users could quickly and easily add common data quality checks on columns and tables. The resulting output of the tests would be saved in the dictionary passed in containing the check names and parameters.

### Use case/motivation

Current SQL Check Operators are extremely broad, but limited to aggregate checks, making them confusing and sometimes redundant with respect for each other. The new checks would allow a more streamlined experience, and would save enough information that the results can be parsed in OpenLineage, which old checks cannot be.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by denimalpaca**
This pull request adds two new SQL Check Operators to the core sql.py file, one called SQLColumnCheckOperator and one called SQLTableCheckOperator. Corresponding tests are added as well. The current functionality of these operators is minimal, but expected to grow over time. These operators, unlike current ones, save information about the results
of the check queries that is then parseable by third party libraries.

closes: #23874

### CODE REVIEW

File: airflow/operators/sql.py [hunk 1]
```diff
@@ -467,6 +467,218 @@ def push(self, meta_data):
...
+def _get_failed_tests(checks):
+    failed_tests = []
+    for check, check_values in checks.items():
+        if not check_values["success"]:
+            failed_tests.append(
+                f"\tCheck: {check}, "
+                f"Pass Value: {check_values['pass_value']}, "
+                f"Result: {check_values['result']}\n"
+            )
+    return failed_tests
```
Comment (uranusjr) [inline airflow/operators/sql.py]:
```suggestion
def _get_failed_tests(checks):
    return [
        f"\tCheck: {check}, "
        f"Pass Value: {check_values['pass_value']}, "
        f"Result: {check_values['result']}\n"
        for check, check_values in checks.items()
        if not check_values["success"]
    ]
```

Comment (uranusjr) [reply airflow/operators/sql.py]: But this string does not seem to be used at all. Maybe it should simply be like this?

```python
def _all_success(checks):
    return all(values["success"] for values in checks.values())
```

Comment (denimalpaca) [reply airflow/operators/sql.py]: The string is used in line 575 in the `AirflowException`. Great suggestion using the list comprehension, thank you!

**Comment by denimalpaca** [comment]
Question I have for reviewers: would it make sense to add the Snowflake, Bigquery, Redshift, etc... versions of this operator in with this PR, or open a separate issue & PR for the provider updates?

**Comment by josh-fell** [comment]
> Question I have for reviewers: would it make sense to add the Snowflake, Bigquery, Redshift, etc... versions of this operator in with this PR, or open a separate issue & PR for the provider updates?

My vote would be for separate ones.

File: airflow/operators/sql.py [hunk 2]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+                'leq_than': 10,
+                'tolerance': 0.2,
+            },
+            'max': {
+                'less_than': 1000,
+                'geq_than': 10,
+                'tolerance': 0.01
+            }
+        }
+    }
```
Comment (josh-fell) [inline airflow/operators/sql.py]: For these pesky build docs errors, you can try adding a `.. code-block::` directive and house these example dicts under them. Something like:
```rst
:param table: the table to run checks on.
:param column_mapping: the dictionary of columns and their associated checks, e.g.:

    .. code-block::

        column_mapping = {
            'col_name': {
                'null_check': {
                    'pass_value': 0,
                },
                'min': {
                    'pass_value': 5,
                    'tolerance': 0.2,
                }
            }
        }

:param conn_id: the connection ID used to connect to the database.
:param database: name of database which overwrite the defined one in connection
```
Check out the docstrings for [`chain()`](https://github.com/apache/airflow/blob/fd4e344880b505cfe53a97a6373d329515bbc7a3/airflow/models/baseoperator.py#L1520) and [`cross_downstream()`](https://github.com/apache/airflow/blob/fd4e344880b505cfe53a97a6373d329515bbc7a3/airflow/models/baseoperator.py#L1649) for examples.

File: airflow/operators/sql.py [hunk 3]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+    :param conn_id: the connection ID used to connect to the database.
+    :param database: name of database which overwrite the defined one in connection
+    """
+
+    column_checks = {
+        # pass value should be number of acceptable nulls
+        "null_check": "SUM(CASE WHEN 'column' IS NULL THEN 1 ELSE 0 END) AS column_null_check",
```
Comment (josh-fell) [inline airflow/operators/sql.py]:
```suggestion
        "null_check": "SUM(CASE WHEN column IS NULL THEN 1 ELSE 0 END) AS column_null_check",
```
Should be checking for the column value rather than the string of the the column name I think?

File: airflow/operators/sql.py [hunk 4]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+
+            if not records:
+                raise AirflowException("The query returned None")
+
+            for idx, result in enumerate(records):
+                tolerance = (
+                    self.column_mapping[column][checks[idx]]["tolerance"]
+                    if "tolerance" in self.column_mapping[column][checks[idx]]
+                    else None
+                )
```
Comment (josh-fell) [inline airflow/operators/sql.py]:
```suggestion
                tolerance = self.column_mapping[column][checks[idx]].get("tolerance")
```

File: airflow/operators/sql.py [hunk 5]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+        if (
+            "greater_than" not in check_values
+            and "geq_than" not in check_values
+            and "less_than" not in check_values
+            and "leq_than" not in check_values
+            and "equal_to" not in check_values
+        ):
+            raise ValueError(
+                "Please provide one or more of: less_than, leq_than, "
+                "greater_than, geq_than, or equal_to in the check's dict."
```
Comment (josh-fell) [inline airflow/operators/sql.py:611]: I think it would be super helpful if listing these equality check options were in the docstring so they can appear in the Python API documentation for users.

File: airflow/operators/sql.py [hunk 6]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+        **kwargs,
+    ):
+        super().__init__(conn_id=conn_id, database=database, **kwargs)
+        for checks in column_mapping.values():
+            for check, check_values in checks.items():
+                self._column_mapping_validation(check, check_values)
+
+        self.table = table
+        self.column_mapping = column_mapping
+        self.sql = f"SELECT * FROM {self.table};"
```
Comment (josh-fell) [inline airflow/operators/sql.py:538]: Just curious, why is this needed here?

File: airflow/operators/sql.py [hunk 7]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+                )
+
+                self.column_mapping[column][checks[idx]]["result"] = result
+                self.column_mapping[column][checks[idx]]["success"] = self._get_match(
+                    self.column_mapping[column][checks[idx]], result, tolerance
+                )
+
+            failed_tests.extend(_get_failed_tests(self.column_mapping[column]))
+        if failed_tests:
+            raise AirflowException(
```
Comment (josh-fell) [inline airflow/operators/sql.py:565]: Maybe it's worth thinking about having a separate exception for Airflow checks/test especially if more data quality functionality is coming?

File: airflow/operators/sql.py [hunk 8]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+        ):
+            raise ValueError(
+                "Please provide one or more of: less_than, leq_than, "
+                "greater_than, geq_than, or equal_to in the check's dict."
+            )
+
+        if "greater_than" in check_values and "less_than" in check_values:
+            if check_values["greater_than"] >= check_values["less_than"]:
+                raise ValueError(
+                    "greater_than should be strictly less than to "
```
Comment (josh-fell) [inline airflow/operators/sql.py]:
```suggestion
                    "greater_than should be strictly less than "
```

File: airflow/operators/sql.py [hunk 9]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+        checks: Dict[str, Dict[str, Any]],
+        conn_id: Optional[str] = None,
+        database: Optional[str] = None,
+        **kwargs,
+    ):
+        super().__init__(conn_id=conn_id, database=database, **kwargs)
+
+        self.table = table
+        self.checks = checks
+        self.sql = f"SELECT * FROM {self.table};"
```
Comment (josh-fell) [inline airflow/operators/sql.py:695]: Same question here as above.

File: airflow/operators/sql.py [hunk 10]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+        for column in self.column_mapping:
+            checks = [*self.column_mapping[column]]
+            checks_sql = ",".join([self.column_checks[check].replace("column", column) for check in checks])
+
+            self.sql = f"SELECT {checks_sql} FROM {self.table};"
+            records = hook.get_first(self.sql)
+            self.log.info(f"Record: {records}")
+
+            if not records:
+                raise AirflowException("The query returned None")
```
Comment (josh-fell) [inline airflow/operators/sql.py]:
```suggestion
            if not records:
                raise AirflowException("The query returned None")

            self.log.info(f"Record: {records}")
```
Seems like it's not worth logging the record info if no records are returned and a failure will occur. WDYT about flipping these around?

Also, WDYT about changing the exception message to something like?
"The following query returned no rows: {sql}"

File: airflow/operators/sql.py [hunk 11]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+        self.sql = f"SELECT * FROM {self.table};"
+
+    def execute(self, context=None):
+        hook = self.get_db_hook()
+        failed_tests = []
+        for column in self.column_mapping:
+            checks = [*self.column_mapping[column]]
+            checks_sql = ",".join([self.column_checks[check].replace("column", column) for check in checks])
+
+            self.sql = f"SELECT {checks_sql} FROM {self.table};"
```
Comment (josh-fell) [inline airflow/operators/sql.py:547]: Similar to the question above, does this need to be an instance attr?

File: airflow/operators/sql.py [hunk 12]
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
...
+            ]
+        )
+
+        self.sql = f"SELECT {checks_sql} FROM {self.table};"
+        records = hook.get_first(self.sql)
+
+        self.log.info(f"Record: {records}")
+
+        if not records:
+            raise AirflowException("The query returned None")
```
Comment (josh-fell) [inline airflow/operators/sql.py]: Same comment as above re: flipping these around.

File: tests/operators/test_sql.py [hunk 13]
```diff
@@ -385,6 +387,82 @@ def test_fail_min_sql_max_value(self, mock_get_db_hook):
             operator.execute()
 
 
+class TestColumnCheckOperator(unittest.TestCase):
```
Comment (josh-fell) [inline tests/operators/test_sql.py:390]: IMO this can be a `pytest` test rather than continuing to use `unittest`.

File: tests/operators/test_sql.py [hunk 14]
```diff
@@ -385,6 +387,82 @@ def test_fail_min_sql_max_value(self, mock_get_db_hook):
...
+            "max": {"less_than": 20, "greater_than": 10},
+        }
+    }
+
+    invalid_column_mapping = {"Y": {"invalid_check_name": {"expectation": 5}}}
+
+    def _construct_operator(self, column_mapping):
+        return SQLColumnCheckOperator(task_id="test_task", table="test_table", column_mapping=column_mapping)
+
+    @mock.patch.object(SQLColumnCheckOperator, "get_db_hook")
```
Comment (josh-fell) [inline tests/operators/test_sql.py:407]: Since every test case uses the same mocked object, you could think about using a pytest fixture for mocking `get_db_hook`. I don't have a strong opinion though.

File: tests/operators/test_sql.py [hunk 15]
```diff
@@ -385,6 +387,82 @@ def test_fail_min_sql_max_value(self, mock_get_db_hook):
...
+        operator = self._construct_operator(self.valid_column_mapping)
+        operator.execute()
+
+    @mock.patch.object(SQLColumnCheckOperator, "get_db_hook")
+    def test_pass_all_checks_inexact_check(self, mock_get_db_hook):
+        mock_hook = mock.Mock()
+        mock_hook.get_first.return_value = (0, 9, 12, 0, 15)
+        mock_get_db_hook.return_value = mock_hook
+        operator = self._construct_operator(self.valid_column_mapping)
+        operator.execute()
```
Comment (josh-fell) [inline tests/operators/test_sql.py:426]: What is the assertion being made in these happy-path tests?

File: airflow/operators/sql.py (hunk 3 above)
Comment (denimalpaca) [reply airflow/operators/sql.py]: Ummm... good question. I'm not 100% sure. Testing on Snowflake, the lack of `'` version had a parse error iirc. So it might be a dialect issue that can be resolved in the provider versions.

File: airflow/operators/sql.py (hunk 6 above)
Comment (denimalpaca) [reply airflow/operators/sql.py:538]: For OpenLineage. The Extractor needs some static query to render, unless we want to have the entire templated query rendered in `__init__()`. I can double-check this, though, it might have changed with the move to the Listener.

File: airflow/operators/sql.py (hunk 10 above)
Comment (denimalpaca) [reply airflow/operators/sql.py]: Ah, good points. Yeah, I think that would be a better order and message.

File: airflow/operators/sql.py (hunk 7 above)
Comment (denimalpaca) [reply airflow/operators/sql.py:565]: I think that's a good idea, but we'd also have to replace exceptions in other SQL operators and I think that's outside the scope of this PR. But I can open an issue and implement that in a different PR. Something like an `AirflowDataQualityException`?

File: tests/operators/test_sql.py (hunk 13 above)
Comment (denimalpaca) [reply tests/operators/test_sql.py:390]: Is there a reason to use one over the other? I was just copying the test pattern from the other ones.

File: tests/operators/test_sql.py (hunk 15 above)
Comment (denimalpaca) [reply tests/operators/test_sql.py:426]: It's implicitly `assert mock_hook.get_first.return_value[i] == self.valid_column_mapping[condition_i]`

File: tests/operators/test_sql.py (hunk 13 above)
Comment (potiuk) [reply tests/operators/test_sql.py:390]: We prefer pytest test type unless there are good reasons not to (for new test)

File: airflow/operators/sql.py (hunk 6 above)
Comment (potiuk) [reply airflow/operators/sql.py:538]: Just to reiterate - those "common" SQL classes will not be released unti Airflow 2.4 and it means that the first operator tha will be able to use them as "mandatory" will be 12 months after 2.4 is released (that's our policy for providers), so I just want to make you aware @denimalpaca that those classes won't be to useful for quite some time. 

Maybe a better Idea will be to add a separate "base sql" provider which will not be in the "airflow/operators" but it will be a separate package that could be installed additionally to airflow. We have not done so before, but maybe there is a good reason to introduce such a "shared sql" provider that wil encapsulate all SQL-like base functionality that open lineage might base on? I think trying to implement it in Airflow core is a bad idea, if the goal is fast adoption by multiple providers. 

Just saying.

Comment (josh-fell) [reply airflow/operators/sql.py:538]: >Maybe a better Idea will be to add a separate "base sql" provider

I think this is an interesting idea and I support it. We could start with these operators and slowly move existing core SQL operators over too.

@potiuk Do you think this provider would be included with Airflow installs like HTTP, FTP, SQLite, and IMAP, or completely separate?

File: tests/operators/test_sql.py (hunk 14 above)
Comment (uranusjr) [reply tests/operators/test_sql.py:407]: If you use pytest fixtures, the `monkeypatch` fixture makes this easy to do in `setup_method`.
